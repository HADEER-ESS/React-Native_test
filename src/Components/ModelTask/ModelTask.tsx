import React, { useState } from 'react'
import { FlatList, Modal, Text, TextInput, TouchableOpacity, View } from 'react-native';
import Todo from '../../Class/TodoClass';
import categoryData from '../../Data/CategoryData';
import CategoryCard from '../categoryCard/CategoryCard';

type TodoItem = {
    id: number,
    title: string,
    description: string
}

const ModelTask = (props: { openModelState: boolean, closeModelFun: React.Dispatch<React.SetStateAction<boolean>> }) => {
    const [taskItem, setTaskItem] = useState<string>('')
    const [taskType, setTaskType] = useState<string>('')
    let [errorMessage, setErrorMessage] = useState<string>("")
    let todoService = Todo.getInstance()

    const handleModelTask = (): void => {
        if (taskItem.length === 0) {
            setErrorMessage("Please Add the task")
        } else {
            setTaskItem('')
            setTaskType('')
            props.closeModelFun(false)
            let item: TodoItem = { id: todoService.taskCount() + 1, title: taskType, description: taskItem }
            todoService.pushToArray(item)

        }
    }

    return (
        <Modal
            visible={props.openModelState}
            animationType="slide"
            transparent={true}
        >
            <TouchableOpacity onPress={() => props.closeModelFun(false)} style={{ flex: 1, justifyContent: 'center', alignItems: 'center', backgroundColor: '#0000000f' }}>
                <View style={{ padding: 16, borderRadius: 11, width: '80%', height: '50%', alignItems: 'center', backgroundColor: '#ffffff' }}>
                    <TextInput
                        placeholder='Enter Todo'
                        testID='Task_input'
                        value={taskItem}
                        style={{ width: '100%', borderRadius: 7, borderColor: "#d6d6d6", paddingHorizontal: 11, alignSelf: 'center', borderWidth: 1 }}
                        onChangeText={txt => setTaskItem(txt)}
                    />
                    <FlatList
                        data={categoryData}
                        horizontal
                        renderItem={({ item }) => <CategoryCard selectedItem={taskType} getTaskType={setTaskType} category={item} />}
                        keyExtractor={item => item.color}
                    />

                    <TouchableOpacity testID='add_btn' style={{ borderRadius: 10, marginVertical: 16, backgroundColor: '#e6e6', alignItems: 'center', width: 70, paddingVertical: 11, alignSelf: 'center' }} onPress={handleModelTask}>
                        <Text style={{ color: '#ffffff', fontSize: 15, fontWeight: '500' }}>Add</Text>
                    </TouchableOpacity>

                    <Text testID='Invalid_text_task'>{errorMessage}</Text>

                </View>
            </TouchableOpacity>
        </Modal>
    )
}

export default ModelTask