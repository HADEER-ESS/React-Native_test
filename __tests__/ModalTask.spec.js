import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native"
import ModelTask from "../src/Components/ModelTask/ModelTask";

//Mock ToDo class
class TODO {
    constructor(title, desc) {
        TODO.id = 1
        this.title = title
        this.desc = desc
    }

    getTodoItem() {
        return { id: TODO.id, title: this.title, desc: this.desc }
    }
}



describe("Test ModalTask Component", () => {

    test("Default Model Task rendering", () => {

        const root = render(<ModelTask />);

        // getAllByPlaceholderText("Enter Todo")
        expect(root).toMatchSnapshot()

    })

    test('updates state and displays input text when user types and presses button', () => {

        const closeModelFun = jest.fn()

        const tasksState = jest.fn()

        const { getByTestId } = render(<ModelTask closeModelFun={closeModelFun} tasksState={tasksState} />)


        const input = getByTestId("Task_input")

        const button = getByTestId("add_btn")

        //simulate user type in textInput
        fireEvent.changeText(input, "New Task")

        expect(input.props.value).toBe("New Task")


        //Simulate button press
        fireEvent.press(button)

        expect(closeModelFun).toHaveBeenCalledWith(false);
        expect(tasksState).toHaveBeenCalledWith({ id: 1, title: '', description: 'New Task' });

    })

    test('displays error message when taskItem is empty and button is pressed', () => {

        const closeModelFun = jest.fn()

        const tasksState = jest.fn()

        const { getByTestId } = render(<ModelTask closeModelFun={closeModelFun} tasksState={tasksState} />)

        const button = getByTestId("add_btn")

        fireEvent.press(button);

        expect(getByTestId("Invalid_text_task")).toHaveTextContent("Please Add a task")

        expect(closeModelFun).not.toHaveBeenCalled()

        expect(tasksState).not.toHaveBeenCalled()
    })
})

