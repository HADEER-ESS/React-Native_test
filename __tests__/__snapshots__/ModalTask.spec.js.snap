// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ModalTask Component Default Model Task rendering 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#0000000f",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderRadius": 11,
          "height": "50%",
          "padding": 16,
          "width": "80%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter Todo"
        style={
          {
            "alignSelf": "center",
            "borderColor": "#d6d6d6",
            "borderRadius": 7,
            "borderWidth": 1,
            "paddingHorizontal": 11,
            "width": "100%",
          }
        }
        testID="Task_input"
        value=""
      />
      <RCTScrollView
        data={
          [
            {
              "color": "red",
              "type": "Urgent",
            },
            {
              "color": "green",
              "type": "Importnat",
            },
            {
              "color": "blue",
              "type": "Upcoming",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={true}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              [
                {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 14,
                  "flexDirection": "row",
                  "height": "15%",
                  "marginHorizontal": 4,
                  "marginTop": 16,
                  "opacity": 1,
                  "paddingHorizontal": 4,
                }
              }
            >
              <View
                style={
                  {
                    "backgroundColor": "red",
                    "borderRadius": 20,
                    "height": 16,
                    "width": 16,
                  }
                }
              />
              <Text
                style={
                  {
                    "color": "#000000",
                    "paddingHorizontal": 3,
                  }
                }
              >
                Urgent
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              [
                {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 14,
                  "flexDirection": "row",
                  "height": "15%",
                  "marginHorizontal": 4,
                  "marginTop": 16,
                  "opacity": 1,
                  "paddingHorizontal": 4,
                }
              }
            >
              <View
                style={
                  {
                    "backgroundColor": "green",
                    "borderRadius": 20,
                    "height": 16,
                    "width": 16,
                  }
                }
              />
              <Text
                style={
                  {
                    "color": "#000000",
                    "paddingHorizontal": 3,
                  }
                }
              >
                Importnat
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              [
                {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 14,
                  "flexDirection": "row",
                  "height": "15%",
                  "marginHorizontal": 4,
                  "marginTop": 16,
                  "opacity": 1,
                  "paddingHorizontal": 4,
                }
              }
            >
              <View
                style={
                  {
                    "backgroundColor": "blue",
                    "borderRadius": 20,
                    "height": 16,
                    "width": 16,
                  }
                }
              />
              <Text
                style={
                  {
                    "color": "#000000",
                    "paddingHorizontal": 3,
                  }
                }
              >
                Upcoming
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#e6e6",
            "borderRadius": 10,
            "marginVertical": 16,
            "opacity": 1,
            "paddingVertical": 11,
            "width": 70,
          }
        }
        testID="add_btn"
      >
        <Text
          style={
            {
              "color": "#ffffff",
              "fontSize": 15,
              "fontWeight": "500",
            }
          }
        >
          Add
        </Text>
      </View>
      <Text
        testID="Invalid_text_task"
      />
    </View>
  </View>
</Modal>
`;
